File: 프로그래머스/3/42895. N으로 표현/N으로 표현.cpp
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

int solution(int N, int number) {
    unordered_set<int> s[9];
    
    int x = N;
    for(int i = 1; i < 9; i++){
        s[i].insert(x);
        x = x * 10 + N;
    }
    
    for(int i = 2; i < 9; i++){
        for(int j = 1; j < i; j++){
            for(int a : s[j]) {
                for(int b : s[i-j]) {
                    s[i].insert(a+b);
                    s[i].insert(a*b);
                    s[i].insert(a-b);
                    if(b != 0) s[i].insert(a/b);
                }
            }
        }
    }
    
    for(int i = 1; i < 9; i++){
        if(s[i].find(number) != s[i].end()) return i;
    }
    return -1;
}

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        List<Set<Integer>> set = new ArrayList<>();
        
        int x = N;
        
        set.add(new HashSet<>());
        
        for(int i = 1; i < 9; i++) {
            set.add(new HashSet<>());
            set.get(i).add(x);
            x = x * 10 + N;
        }
        
        for(int i = 2; i < 9; i++){
            for(int j = 1; j < i; j++){
                for(int a : set.get(j)) {
                    for(int b : set.get(i-j)) {
                        set.get(i).add(a + b);
                        set.get(i).add(a * b);
                        set.get(i).add(a - b);
                        if(b != 0) set.get(i).add(a / b);
                    }
                }
            }
        }

        for(int i = 1; i < 9; i++){
            if(set.get(i).contains(number)) return i;
        }
        return -1;

    }
}